File: src/main/kotlin/data/DangeonMap.java
package data;

import data.model.*;
import data.repository.Entity;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class DangeonMap {
    private final int width;
    private final int height;
    private final Player player;
    private final List<Entity> entities = new ArrayList<>();
    private int enemiesCount = 0;
    private final Random random = new Random();

    public DangeonMap(int width, int height, Player player) {
        this.width = width;
        this.height = height;
        this.player = player;
    }

    public int getEnemiesCount() { return enemiesCount; }

    public boolean isInside(int x, int y) {
        return x >= 0 && x < width && y >= 0 && y < height;
    }

    public Entity getEntityAt(int x, int y) {
        for (Entity e : entities) {
            if (e.getX() == x && e.getY() == y) return e;
        }
        return null;
    }

    public boolean isFree(int x, int y) {
        return isInside(x, y) && getEntityAt(x, y) == null && (player.getX() != x || player.getY() != y);
    }

    public void render() {
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                if (player.getX() == x && player.getY() == y) System.out.print("@ ");
                else {
                    Entity e = getEntityAt(x, y);
                    System.out.print((e != null ? e.getSymbol() : '.') + " ");
                }
            }
            System.out.println();
        }
        System.out.println("HP: " + player.getHp() + " | В инвентаре " + player.getInventory().size() + " предметов");
    }

    public void spawnEnemies(int count) {
        enemiesCount = count;
        for (int i=0;i<count;i++) {
            int[] coords = randomFreeCell();
            entities.add(new Enemy(coords[0], coords[1]));
        }
    }

    public void spawnItems(int count) {
        for (int i=0;i<count;i++) {
            int[] coords = randomFreeCell();
            entities.add(ItemFactory.createRandomItem(coords[0], coords[1]));
        }
    }

    public void movePlayer(int dx, int dy) {
        int newX = player.getX() + dx;
        int newY = player.getY() + dy;
        if (!isInside(newX, newY)) {
            System.out.println("Нельзя ходить в стену!");
            return;
        }

        Entity entity = getEntityAt(newX, newY);

        if (entity instanceof Enemy) {
            Enemy enemy = (Enemy) entity;
            player.attack(enemy);
            if (enemy.isAlive()) enemy.attack(player);
            else {
                System.out.println("Враг побеждён");
                enemiesCount--;
                entities.remove(enemy);
            }
        } else if (entity instanceof Item) {
            Item item = (Item) entity;
            System.out.println("Ты нашёл " + item.getVisibleName(player) + "! Взять? (y/n)");
            String input = System.console() != null ? System.console().readLine().trim().toLowerCase() : "y"; // безопасно для IDE
            if (input.equals("y")) {
                player.getInventory().add(item);
                entities.remove(item);
            }
            player.setX(newX);
            player.setY(newY);
        } else {
            player.setX(newX);
            player.setY(newY);
        }
    }

    private int[] randomFreeCell() {
        int x, y;
        do {
            x = random.nextInt(width);
            y = random.nextInt(height);
        } while (!isFree(x, y));
        return new int[]{x, y};
    }
}

File: src/main/kotlin/data/DangeonMap.kt
package ru.kaed.data

import ru.kaed.data.model.Enemy
import ru.kaed.data.model.Item
import ru.kaed.data.model.ItemEffect
import ru.kaed.data.model.Player
import ru.kaed.data.repository.Entity
import kotlin.math.min
import kotlin.random.Random

data class DungeonMap(
    val width: Int,
    val height: Int,
    val player: Player,
    val entities: MutableList<Entity> = mutableListOf()
) {
    fun isInside(x: Int, y: Int) = x in 0..<width && y in 0..<height

    fun getEntityAt(x: Int, y: Int): Entity? = entities.find { it.x == x && it.y == y }

    fun isFree(x: Int, y: Int): Boolean =
        isInside(x, y) && getEntityAt(x, y) == null && (player.x != x || player.y != y)

    fun render() {
        for (y in 0..<height) {
            for (x in 0..<width) {
                when {
                    player.x == x && player.y == y -> print("@ ")
                    else -> print((getEntityAt(x, y)?.symbol ?: '.') + " ")
                }
            }
            println()
        }
        println("HP: ${player.hp} | В инвентаре ${player.inventory.size} предметов")
    }


    fun spawnEnemies(count: Int) {
        repeat(count) {
            val (x, y) = randomFreeCell()
            entities.add(Enemy(x, y))
        }
    }

    fun spawnItems(count: Int) {
        repeat(count) {
            val (x, y) = randomFreeCell()
            entities.add(ItemFactory.createRandomItem(x, y, player))
        }
    }

    fun movePlayer(dx: Int, dy: Int){
        val newX = player.x + dx
        val newY = player.y + dy
        if (!isInside(newX, newY)) {
            println("Нельзя ходить в стену!")
            return
        }

        val entity = getEntityAt(newX, newY)

        when(entity){
            is Enemy -> {
                player.attack(entity)
                if (entity.isAlive()) entity.attack(player)
                else {
                    println("Враг побеждён")
                    entities.remove(entity)
                }
            }
            is Item -> {
                println("Ты нашёл ${entity.name}! Взять? (y/n)")
                if (readln().trim().lowercase() == "y") {
                    player.inventory.add(entity)
                    entities.remove(entity)
                }
                player.x = newX
                player.y = newY
            }
            else -> {
                player.x = newX
                player.y = newY
            }
        }
    }

    private fun randomFreeCell(): Pair<Int, Int> {
        var x: Int
        var y: Int
        do {
            x = Random.nextInt(width)
            y = Random.nextInt(height)
        } while (!isFree(x, y))
        return x to y
    }

}

object ItemFactory {
    fun createRandomItem(x: Int, y: Int, player: Player): Item {
        val roll = Random.nextInt(100)

        val effect: ItemEffect = when {
            roll < 50 -> ItemEffect.Heal(Random.nextInt(3, 11))
            roll < 75 -> ItemEffect.BuffAttack(Random.nextInt(3, 11))
            roll < 90 -> ItemEffect.Poison(Random.nextInt(2, 6))
            else -> ItemEffect.DebuffAttack(Random.nextInt(1, 4))
        }

        return Item(
            x = x,
            y = y,
            name = effect.toString(),
            effect = applyEffect(player = player, effect = effect)
        )
    }

    private fun applyEffect(player: Player, effect: ItemEffect): (Player) -> Unit = {
        when (effect) {
            is ItemEffect.Heal -> {
                val currentHP = min(effect.amount + player.hp, 3)
                val increasedHP = currentHP - player.hp
                player.hp = currentHP
                println("Здоровье восстановлено на ${increasedHP}, HP: ${player.hp}")
            }
            is ItemEffect.Poison -> {
                player.hp -= effect.amount
                println("Здоровье уменьшено на ${effect.amount}, HP: ${player.hp}")
            }
            is ItemEffect.BuffAttack -> {
                player.attackPower += effect.amount
                println("Атака увеличена на ${effect.amount}, теперь ты наносишь ${player.attackPower} урона")
            }
            is ItemEffect.DebuffAttack -> {
                player.attackPower -= effect.amount
                println("Атака Уменьшена на ${effect.amount}, теперь ты наносишь ${player.attackPower} урона")
            }
        }
    }


}